pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// KongCustomEntityList is a list of kong custom entities.
+type KongCustomEntityList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongCustomEntity `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// Fields is the list of fields in the entity.
+	Fields apiextensionsv1.JSON `json:"fields"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=kongcustomentitydefinitions,shortName=kced,categories=kong-ingress-controller
+
+// KongCustomEntityDefinition represents definition of a custom entity type in Kong.
+type KongCustomEntityDefinition struct {
```

Comment:
Same as for `KongCustomEntity`, I'd suggest adding a separate `KongCustomEntitySpec` and put `Name`, `AdminAPIName`, and `Dependencies` into it.

As for the conditions, I think `Programmed` doesn't make sense for this one as it's not really used to populate anything in Kong. It's only metadata that's needed to call proper endpoints when running in DB mode and can probably be shared between controllers fine.

Reply:
I think `KongCustomEntityDefinition` is similar to definition of CRD itself, so this resource does not fit the "spec-status" model. In other words, what attributes of  `KongCustomEntityDefinition`  should be put in spec and what should be put in status?

Reply:
Regarding the conditions and ingressClass:

On the other hand, I can imagine that sometime in the future:
- we could come up with some status that we would like to populate for `KongCustomEntityDefinition` (e.g. number of configured entities of a given type)
- we may want to support different versions of a single custom entity (same name, but different AdminAPIName/Dependencies) between multiple controller instances (e.g. one running with Kong version X, the other one with X+1 where a custom entity definition was modified). 

That makes me wonder whether we should also add `ingressClass` to `KongCustomEntityDefinition` as well.

Reply:
Entity definitions can be reused across separate instances fine and are cluster-scoped, so they shouldn't have class info.  You shouldn't have to define that degraphQL routes exist twice and remember which name you need to use for a given install; they're going to be identical internally.

This could technically introduce issues for version skew, but hopefully gateway doesn't change the database names for entities between versions. If they do, we ask them to not do that (it serves no purpose!) and can still begrudgingly create two classless variants of the same entity to account for the differences.

There's no obvious reason to add status to these, but FWIW, IngressClass is the same way and does use a Spec field.

Reply:
moved the current attributes to `spec`.